### PR TITLE
Fix petsc copy

### LINKED ISSUE #4350: Issues in copy_petsc_vector
In commit [88b34fd](https://github.com/dealii/dealii/commit/88b34fd76beaa45fc6192018148c5c1d930e781f), @drwells imported a new function [`copy_petsc_vector`](https://github.com/dealii/dealii/blob/master/include/deal.II/lac/vector.templates.h#L102) to generalize the copying operation from PETSc vector to deal.II `Vector`. However, I found two issues in this function:

1. The custom deleter provided to the `unique_ptr` only destroys the `Vec` object, it does not release the memory allocated by `new`. This may cause memory leaks. Am I right?
2. The MPI communicator provided to `VecCreateSeq` can only contain one process. Calling this function on more that one process we will get the following error messages:
    PETSC ERROR: Invalid argument
    PETSC ERROR: Cannot create VECSEQ on more than one process
Also, VecCopy can not be used to copy distributed vector to sequential vector. Maybe we should use a `VecScatter`? PETSc does provide a function named `VecScatterCreateToAll` to copy all values to each processor.

### PR DESCRIPTION
**Pull request by drwells**
@adamqc Thank you again for catching this. I am impressed by your vigilance.

Closes #4350 by fixing a variety of bugs I created:
1. Use a plain destructor instead of a `unique_ptr` to destroy PETSc classes in case of exceptions
2. Correctly set things up with a vector scattering context

I added a test for this (it works with this patch and release versions of the library but not with `master`).

### CODE REVIEW

**Comment by adamqc** [comment]
Seems ok to me. Closed #4350.

File: doc/news/changes/incompatibilities/20170506DavidWells [hunk 1]
```diff
@@ -0,0 +1,3 @@
+Removed: The deprecated serial PETSc vector classes have been removed.
```
Comment (bangerth) [inline doc/news/changes/incompatibilities/20170506DavidWells]: is this the wrong changelog entry?

**Comment by bangerth** [comment]
As for the `unique_ptr`, wouldn't it have been enough to just add the `delete` call to the custom deleter lambda function?

I suspect the setup of the scatter operation was genuinely wrong before?

File: doc/news/changes/incompatibilities/20170506DavidWells (hunk 1 above)
Comment (drwells) [inline doc/news/changes/incompatibilities/20170506DavidWells]: Technically yes. Philosophically, I consider this PR to fix all of the bugs with #4279, which includes 'not adding a changelog entry'.

**Comment by drwells** [comment]
Yes (I forgot to call the matching `delete` for the `new Vec()`), but now we have to manage two objects, and since nothing leaves the scope we may as well use one destructor instead of multiple `unique_ptrs`.

I do not believe that the previous scatter implementation (i.e., the one that used to be in `petsc_vector.h`) was wrong aside from exception safety issues.

File: doc/news/changes/incompatibilities/20170506DavidWells (hunk 1 above)
Comment (bangerth) [inline doc/news/changes/incompatibilities/20170506DavidWells]: Fair enough ;-)

**Comment by bangerth** [comment]
I'm not fundamentally opposed to this kind of patch, but it seems unnecessarily complicated and hard to read to put all of the destructor logic at the top of the function so that it gets executed at the bottom.

This would make sense if there was a real threat of exceptions happening somewhere in between. But I don't think there is. We only really call C functions that can't throw exceptions, and if something *should* happen after all, all hope is lost anyway because there is no reasonable way to recover. So there is a lot of obfuscation for little tangible benefit -- I would tolerate a memory leak in that case. That's a price we pay for dealing with software written in C.

Secondly, this style isn't used anywhere else. We probably have several dozen places where we create and delete PETSc objects in a temporary way like this, but none of them use this idiom of cleaning up. I think that also makes it hard to read, because it's a one-off.

Not sure what to suggest -- I don't want to stand in the way of a good patch, but I'm also not excited about it :-)

**Comment by drwells** [comment]
I had the same thoughts when writing this regarding our exception safety (i.e., if we crash then this doesn't really matter). I will go back and remove the destructor logic.

**Comment by drwells** [comment]
I updated the patch.

**Comment by bangerth** [comment]
OK if the tester is happy.

**Pull request merged**